ui/src/components/chat/compose/messages.rs

```diff
@@ -591,39 +591,69 @@ struct MessagesProps<'a> {
     pending: bool,
 }
 fn render_messages<'a>(cx: Scope<'a, MessagesProps<'a>>) -> Element<'a> {
-    let state = use_shared_state::<State>(cx)?;
     let edit_msg: &UseState<Option<Uuid>> = use_state(cx, || None);
     // see comment in ContextMenu about this variable.
     let reacting_to: &UseState<Option<Uuid>> = use_state(cx, || None);
-
-    let ch = use_coroutine_handle::<MessagesCommand>(cx)?;
+    let state = use_shared_state::<State>(cx)?;
     cx.render(rsx!(cx.props.messages.iter().map(|grouped_message| {
-        let should_fetch_more = grouped_message.should_fetch_more;
-        let message = &grouped_message.message;
-        let sender_is_self = message.inner.sender() == state.read().did_key();
-
-        // WARNING: these keys are required to prevent a bug with the context menu, which manifests when deleting messages.
-        let is_editing = edit_msg
-            .get()
-            .map(|id| !cx.props.is_remote && (id == message.inner.id()))
-            .unwrap_or(false);
-        let context_key = format!("message-{}-{}", &message.key, is_editing);
-        let message_key = format!("{}-{:?}", &message.key, is_editing);
-        let msg_uuid = message.inner.id();
-        let conversation_id = message.inner.conversation_id();
-
-        if cx.props.pending {
-            return rsx!(render_message {
-                message: grouped_message,
-                is_remote: cx.props.is_remote,
-                message_key: message_key,
-                edit_msg: edit_msg.clone(),
-                pending: cx.props.pending
-            });
-        }
+        cx.render(rsx!(render_single_message {
+            message: grouped_message,
+            active_chat_id: cx.props.active_chat_id,
+            num_messages_in_conversation: cx.props.num_messages_in_conversation,
+            is_remote: cx.props.is_remote,
+            pending: cx.props.pending,
+            edit_msg: edit_msg.clone(),
+            reacting_to: reacting_to.clone(),
+            state: state,
+        }))
+    })))
+}
 
-        // todo: add onblur event
-        rsx!(ContextMenu {
+#[derive(Props)]
+struct MessageContextProps<'a> {
+    message: &'a GroupedMessage<'a>,
```

Comment:
BTW @Demonthos and I were looking at this, and in an upcoming update we're going to disallow internal references here.

Your usecase is not problematic, but there are potential bugs abound with things like `RefCell<'a>` that we can't autoheuristic.

Reply:
Also, @Demonthos has been working on making the local subscriptions implicit by building some of the logic into Dioxus itself.

https://github.com/DioxusLabs/dioxus/pull/1300

That will likely be merged within the 0.4 cycle so you guys can benefit from it without another big version bump.

Reply:
We found a way around breaking internally references. A new minor version of Dioxus core has been published, but it will not break this code